### PR TITLE
Optimizing client

### LINKED ISSUE #4108: Client software could be a lot smarter
#### Problem

On the live network with lots of distributed nodes, a client would want to optimize itself for talking to its closest peers and also actively maintain lists of active peers which have the newest slot and blockhashes and good response time. It might also want to send its transactions to a future leader to get a better chance of the transactions succeeding on the network.

#### Proposed Solution

* Higher-level optimizing client which keeps track of live nodes and response times
* Get slot height with blockhash to tell which nodes have fallen behind
* Think about RPC interface for determining who the next leader is.

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Clients are fixed to a single validator node which may be unavailable and if that's the case the client usually hangs or crashes.

#### Summary of Changes

Pick random validator nodes for RPC, keep some stats on how those nodes responded in the past to pick the best ones.

Fixes #4108

### CODE REVIEW

File: client/src/optimizing_client.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+use solana_sdk::transaction::{self, Transaction};
+use solana_sdk::transport::Result as TransportResult;
+use std::collections::VecDeque;
+use std::io;
+use std::sync::RwLock;
+use std::time::Duration;
+use std::time::Instant;
+
+/// An object for querying and sending transactions to the network.
+pub struct OptimizingClient {
```
Comment (garious) [inline client/src/optimizing_client.rs]: How about calling this something like `ValidatorClients` with one method `closest(&self) -> &ThinClient`?

Comment (sakridge) [reply client/src/optimizing_client.rs]: Well, I think the behavior is randomly testing a set of nodes so closest to me sounds like a misnomer. Isn't it also nice to be able to switch out this with the other clients like BankClient and the regular client?

Comment (garious) [reply client/src/optimizing_client.rs]: It's a nice abstraction, but looks like a leaky one. It doesn't seem like one would want to dynamically switch clients at each and every client call. You'd want to pick a closest client, use it for a bit, and then perhaps switch.  Also, is it worth maintaining a wrapper that needs to be updated any time that trait changes?

Comment (garious) [reply client/src/optimizing_client.rs]: Also, another way to do this is within ThinClient. Instead of swapping clients you can swap socket addresses.

Comment (garious) [reply client/src/optimizing_client.rs]: `client()` instead of `closest()`?

Comment (sakridge) [reply client/src/optimizing_client.rs]: Sure, switching each call might be excessive, but this gives you the flexibility to try any strategy.  Maintaining the interface is a bit annoying, I guess in your suggestion you would have some kind of experimentation period maybe on creation and then periodically, then the user would call this client() method to get the best one.

Comment (sakridge) [reply client/src/optimizing_client.rs]: The client may have better information about when to experiment, then in the shim design you need to give some hints how to operate.

Comment (garious) [reply client/src/optimizing_client.rs]: Yes, exactly. My suggestion is to give the user the control until we start seeing a pattern of every user using it the same way. At that point (which may be never) we formalize it with an abstraction.

Comment (sakridge) [reply client/src/optimizing_client.rs]: I used switching sockets in thin_client.

**Pull request merged**